types/react-redux/index.d.ts

```diff
@@ -171,7 +171,7 @@ export interface Connect {
 
     <no_state = {}, TDispatchProps = {}, TOwnProps = {}>(
         mapStateToProps: null | undefined,
-        mapDispatchToProps: TDispatchProps,
+        mapDispatchToProps: MapDispatchToPropsParam<TDispatchProps, TOwnProps>,
```

Comment:
I took some time and I think I understand your change now.

One question: why is `ResolveThunks` currently used only in the "object" case?
Is there any potential negative effect of using that in the union case (i.e. is it a problem if the type system can't tell whether `mapDispatchToProps` is an object or not, but at runtime it actually is a function, but the `ResolveThunks` forced something else to be a certain way elsewhere)?

Reply:
I'm not sure I can answer that question. I did a `git blame` to find where this was originally added, and I came across https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30255, however I couldn't find any discussion in there about this. /cc @surgeboris who implemented that PR.

Reply:
Note that my point is not that I think it should be used in the non-object case, but rather that I can't tell from documentation whether you would or wouldn't want to. I think that this change would, when providing a function as `mapDispatchToProps`, first try checking without resolving thunks and then try without it.

Reply:
I'm not quite sure I follow. Can you elaborate more on what changes you think might need to be made, or what I might need to investigate before we can get this PR approved? Sorry if I'm being slow here.

Reply:
I looked at it some more. I think what I worried about was a case like the following:
```ts
const simpleAction = (payload: boolean) => ({
        type: 'SIMPLE_ACTION',
        payload,
    });

    interface OwnProps {
        foo: string;
    }
    interface TestComponentProps extends OwnProps {
        simpleAction: typeof simpleAction;
        simpleAction2: typeof simpleAction;
    }
    class TestComponent extends React.Component<TestComponentProps> { }

    const mapStateToProps = ({ foo }: { foo: string }) => ({ foo })

    const mapDispatchToProps2 = (dispatch: Dispatch) => {
        return {
            simpleAction: (payload: boolean) => dispatch(simpleAction(payload)),
            simpleAction2: (payload: boolean) => () => dispatch(simpleAction(payload))
        };
    };

    const Test1 = connect(mapStateToProps, mapDispatchToProps)(TestComponent);
    const verify = <div>
        <Test1 foo='bar' />
    </div>;
```
I was under the impression that the type checker would first pick the `MapDispatchToPropsNonObject` overload and fail to check (because of the type of `simpleAction2`), _and then go back_ and pick the second overload with the new change and succeed because of `ResolveThunks` thinking that the value provided for `simpleAction2` is a thunk. Turns out that's not how the overload resolution works - there's no backtracking like I expected. Do you want to add this as an `// $ExpectError` test? if not, I'll go ahead and merge :) Thanks!

Reply:
> Do you want to add this as an `// $ExpectError` test? if not, I'll go ahead and merge :) Thanks!

Thanks for clarifying. Unfortunately I still don't understand, so I'm hesitant to add the test. Would you like me to add it anyway? Or perhaps that test could be added in a separate PR, if it's not related to this change?

Reply:
That's fine, I'll consider adding it separately.
Thanks for the contribution; sorry I delayed it so long!